src/zalsa.rs

```diff
@@ -502,68 +500,63 @@ where
         if cached_data == Self::UNINITIALIZED {
-            #[cold]
-            #[inline(never)]
-            fn get_or_create_index_slow<I: Ingredient>(
-                this: &IngredientCache<I>,
-                zalsa: &Zalsa,
-                create_index: impl Fn() -> IngredientIndex,
-            ) -> IngredientIndex {
-                let index = create_index();
-                let nonce = zalsa.nonce().into_u32().get() as u64;
-                let packed = (nonce << u32::BITS) | (index.as_u32() as u64);
-                debug_assert_ne!(packed, IngredientCache::<I>::UNINITIALIZED);
-
-                // Discard the result, whether we won over the cache or not does not matter
-                // we know that something has been cached now
-                _ = this.cached_data.compare_exchange(
-                    IngredientCache::<I>::UNINITIALIZED,
-                    packed,
-                    Ordering::Release,
-                    Ordering::Acquire,
-                );
-                // and we already have our index computed so we can just use that
-                index
-            }
-
-            return get_or_create_index_slow(self, zalsa, create_index);
+            return self.get_or_create_index_slow(zalsa, create_index);
         };
 
-        // unpack our u64
-        // SAFETY: We've checked against `UNINITIALIZED` (0) above and so the upper bits must be non-zero
+        // Unpack our `u64` into the nonce and index.
+        let index = IngredientIndex(cached_data as u32);
+
+        // SAFETY: We've checked against `UNINITIALIZED` (0) above and so the upper bits must be non-zero.
         let nonce = Nonce::<StorageNonce>::from_u32(unsafe {
             NonZeroU32::new_unchecked((cached_data >> u32::BITS) as u32)
         });
-        let mut index = IngredientIndex(cached_data as u32);
 
+        // The data was cached for a different database, we have to ensure the ingredient was
+        // created in ours.
         if zalsa.nonce() != nonce {
-            index = create_index();
+            let (_, ingredient) = create_index();
+            return ingredient;
         }
-        index
+
+        zalsa.lookup_ingredient(index).assert_type()
+    }
+
+    #[cold]
+    #[inline(never)]
+    fn get_or_create_index_slow<'db>(
+        &self,
+        zalsa: &'db Zalsa,
+        create_index: impl Fn() -> (IngredientIndex, &'db I),
+    ) -> &'db I {
+        let (index, ingredient) = create_index();
+        let nonce = zalsa.nonce().into_u32().get() as u64;
+        let packed = (nonce << u32::BITS) | (index.as_u32() as u64);
+        debug_assert_ne!(packed, IngredientCache::<I>::UNINITIALIZED);
+
+        // Discard the result, whether we won over the cache or not doesn't matter.
+        _ = self.cached_data.compare_exchange(
+            IngredientCache::<I>::UNINITIALIZED,
+            packed,
+            Ordering::Release,
+            Ordering::Relaxed,
```

Comment:
I relaxed the failure ordering here because we don't synchronize on failure. FWIW this `compare_exchange` could even just be a `store` because it encapsulates all the cached data, but it doesn't really matter.